File: sword_offer/python/05_用两个栈实现队列.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

import time
import random
import StackNode

"""
class stackNode():
    def __init__(self, value=0):
        self.val = value
        self.next = None
"""

class Solution():
    """
    使用两个堆栈实现一个队列
    """
    def __init__(self):
        """
        初始化
        """
        # 使用list封装
        self.s1 = []
        self.s2 = []
        # 使用自定义的栈
        self.stack1 = StackNode.StackNode()
        self.stack2 = StackNode.StackNode()
        
    # ===== 使用list封装 ===== 
    def push(self, val):
        """
        压入
        """
        self.s1.append(val)
    
    def pop(self):
        """
        弹出
        """
        if self.s1 == [] and self.s2 == []:
            return 0
        
        if self.s2 != []:
            return self.s2.pop(-1)
        
        while self.s1:
            self.s2.append(self.s1.pop(-1))
        
        return self.pop()
    
    # ===== 使用自定义的栈 =====
    def push2(self, val):
        """
        压入
        """
        self.stack1.push(val)

    def pop2(self):
        """
        弹出
        """
        if self.stack1.empty() and self.stack2.empty():
            return 0
        
        if not self.stack2.empty():
            tmp = self.stack2.top()
            self.stack2.pop()
            return tmp
        
        while not self.stack1.empty():
            tmp = self.stack1.top()
            self.stack1.pop()
            self.stack2.push(tmp)
        
        return self.pop2()

    def test(self, lst):
        """
        测试函数
        输入lst，全部push，然后全部pop
        """
        print("=====")
        start = time.time()
        for i in lst:
            self.push(i)
        for i in lst:
            print("{:d}, ".format(self.pop()), end="")
        end = time.time()
        print("times(us): {:>5.2f}".format((end - start)*10**6))

        start = time.time()
        for i in lst:
            self.push2(i)
        for i in lst:
            print("{:d}, ".format(self.pop2()), end="")
        end = time.time()
        print("times(us): {:>5.2f}".format((end - start)*10**6))


def main():
    lst = [1, 2, 3, 5, 7]
    lst2 = list(random.sample(range(20), 20))
    lst3 = list(range(100))

    s = Solution()
    s.test(lst)
    s.test(lst2)
    s.test(lst3)


if __name__ == "__main__":
    main()

File: sword_offer/cpp/05_用两个栈实现队列.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <stdio.h>
#include <sys/time.h>

#include "StackNode.h"
#include "my_vector.h"

/*
typedef struct StackNode {
    int val;
    StackNode* next;
    StackNode(int input) : val(input), next(nullptr) {};  
} StackNode;
*/

void printf_stack(std::stack<int>);

class Solution {
public:
    // ===== 使用stl定义的栈 =====
    std::stack<int> s1;
    std::stack<int> s2;

    void push(int node) {
        s1.push(node);
    }

    int pop(void) {
        if (s1.empty() && s2.empty()) {
            return 0;
        }

        if (!s2.empty()) {
            int tmp = s2.top();
            s2.pop();
            return tmp;
        } 

        while (!s1.empty()) {
            s2.push(s1.top());
            s1.pop();
        }
        return pop();
    }

    // ===== 使用自定义的栈 =====
    StackNode* h1;
    StackNode* h2;

    void push2(int node) {
        StackNode_push(h1, node);
    }

    int pop2(void) {
        if (h1 == nullptr && h2 == nullptr) {
            return 0;
        }

        if (h2 != nullptr) {
            int tmp = StackNode_top(h2);
            StackNode_pop(h2);
            return tmp;
        }

        while (h1 != nullptr) {
            StackNode_push(h2, StackNode_top(h1));
            StackNode_pop(h1);
        }
        return pop2();
    }

    // ===== 构造函数以及析构函数 =====
    Solution() {
        h1 = nullptr;
        h2 = nullptr;
    }

    ~Solution() {};

    // ===== 测试函数 =====
    // 输入vec，全部push，然后全部pop
    void test(std::vector<int>& vec) {
        struct timeval start, end;
        
        printf("=====\n");
        gettimeofday(&start, nullptr);
        for (auto & i : vec) {
            push(i);
        }
        for (auto & i : vec) {
            printf("%d, ", pop());
        }
        gettimeofday(&end, nullptr);
        printf("times(us):%ld\n", end.tv_usec - start.tv_usec);

        gettimeofday(&start, nullptr);
        for (auto & i : vec) {
            push2(i);
        }
        for (auto & i : vec) {
            printf("%d, ", pop2());
        }
        gettimeofday(&end, nullptr);
        printf("times(us):%ld\n", end.tv_usec - start.tv_usec);
    }
};

int main(int argc, char* argv[])
{
    std::vector<int> vec = {1, 3, 5, 6, 7};
    std::vector<int> vec2 = vector_CreatRandom(1, 20, 20);
    std::vector<int> vec3 = vector_CreatRange(1, 100, 1);

    Solution s;
    s.test(vec);
    s.test(vec2);
    s.test(vec3);

    return 0;
}